src/plstring.c

```diff
@@ -0,0 +1,540 @@
+#include <nitro.h>
+#include <string.h>
+
+#include "struct_decls/struct_plstring_decl.h"
+
+#include "heap.h"
+#include "plstring.h"
+
+// Important character constants
+#define CR                      0xe000
+#define TERMINATOR              0xffff
+#define IS_COMPRESSED_MASK      0xf100
+#define COMPRESSION_MASK        0x01ff
+#define COMPRESSED_TERMINATOR   0x01ff
+
+// Allocation checksum + check function
+#define IS_ALLOCED_MASK         0xb6f8d2ec
+#define PLSTRING_CHECK(string) { \
+    GF_ASSERT(string != NULL); \
+    GF_ASSERT(string->checksum == IS_ALLOCED_MASK); \
+}
+
+/**
+ * Structure definition for a managed string buffer.
+ */
+struct PLString_t {
+    u16 size;               ///< Maximum number of characters that this buffer will support.
+
+    u16 len;                ///< Length of this string's content, i.e. number of characters
+                            ///< actually present in the string (as opposed to many it can
+                            ///< hold).
+
+    u32 checksum;           ///< Magic number checksum assigned at allocation time.
+                            ///< This prop is changed to a different value when freeing the
+                            ///< buffer in order to prevent further operations.
+
+    PLChar * buffer;        ///< The actual string content, with each character encoded as
+                            ///< a pair of bytes.
+};
+
+#define PLSTRING_HEADER_SIZE (sizeof(PLString) - sizeof(PLChar))
+
+PLString * PLString_Init (u32 size, u32 heapID)
+{
+    PLString * newString = Heap_AllocFromHeap(
+        heapID,
+        PLSTRING_HEADER_SIZE + (sizeof(PLCHAR) * size)
+    );
+
+    if (newString) {
+        newString->size         = size;
+        newString->len          = 0;
+        newString->checksum     = IS_ALLOCED_MASK;
+        newString->buffer[0]    = TERMINATOR;
+    }
+
+    return newString;
+}
+
+void PLString_Free (PLString * string)
+{
+    PLSTRING_CHECK(string);
+
+    string->checksum = IS_ALLOCED_MASK + 1;
+    Heap_FreeToHeap(string);
+}
+
+void PLString_Clear (PLString * string)
+{
+    PLSTRING_CHECK(string);
+
+    string->len         = 0;
+    string->buffer[0]   = TERMINATOR;
+}
+
+void PLString_Copy (PLString * dst, const PLString * src)
+{
+    PLSTRING_CHECK(dst);
+    PLSTRING_CHECK(src);
+
+    if (dst->size > src->len) {
+        memcpy(dst->buffer, src->buffer, (src->len + 1) * sizeof(PLChar));
+        dst->len = src->len;
+
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+PLString * PLString_Clone (const PLString * src, u32 heapID)
+{
+    PLSTRING_CHECK(src);
+
+    PLString * dst = PLString_Init(src->len + 1, heapID);
+
+    if (dst) {
+        PLString_Copy(dst, src);
+    }
+
+    return dst;
+}
+
+void PLString_FromInt (PLString * dst, int num, u32 len, NUMBER_PADDING_MODE paddingMode, CHARACTER_WIDTH charWidth)
```

Comment:
`typedef enum` is not allowed. Enum types should be PascalCase. Use `enum MyEnum` when referencing the type in source code.